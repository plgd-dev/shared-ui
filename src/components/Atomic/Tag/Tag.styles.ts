import { css } from '@emotion/react'
import { colors } from '../_utils/colors'

export const tag = css`
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    background: #23233c;
    border-radius: 4px;
    font-family: 'Poppins', sans-serif;
    font-style: normal;
    font-weight: 400;
    font-size: 12px;
    line-height: 18px;
    color: ${colors.neutral500};
    padding: 3px 8px;
`

export const clickable = css`
    cursor: pointer;
`

export const icon = css`
    margin-right: 4px;
`

export const blue = css`
    color: ${colors.primary};
    background: rgba(135, 206, 242, 0.2);
`

export const white = css`
    color: ${colors.neutral500};
    background: #fff;
`
