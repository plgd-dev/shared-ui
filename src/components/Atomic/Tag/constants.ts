export const tagVariants = {
    DEFAULT: 'default',
    BLUE: 'blue',
    WHITE: 'white',
} as const
