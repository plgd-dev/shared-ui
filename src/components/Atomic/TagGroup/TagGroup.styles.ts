import { css } from '@emotion/react'

export const parent = css`
    width: 100%;
    position: relative;
`

export const tagGroup = css`
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-wrap: wrap;
    gap: 4px;
`

export const justifyEnd = css`
    justify-content: end;
`

export const tag = css``

export const test = css`
    position: absolute;
    width: 1px;
    height: 1px;
    visibility: hidden;
`

export const testInner = css`
    display: flex;
`
