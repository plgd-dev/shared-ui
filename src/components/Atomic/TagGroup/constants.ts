export const justifyContent = {
    START: 'start',
    END: 'end',
} as const
