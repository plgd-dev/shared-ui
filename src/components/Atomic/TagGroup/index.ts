export { default } from './TagGroup'
export * from './TagGroup'

export * from './constants'
