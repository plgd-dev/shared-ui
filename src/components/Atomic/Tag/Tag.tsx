import { FC } from 'react'
import { defaultProps, Props } from './Tag.types'
import * as styles from './Tag.styles'
import { Icon } from '../Icon'
import isFunction from 'lodash/isFunction'
import { tagVariants } from './constants'

const Tag: FC<Props> = (props) => {
    const { children, className, icon, id, onClick, variant } = { ...defaultProps, ...props }
    return (
        <div
            className={className}
            css={[styles.tag, isFunction(onClick) && styles.clickable, variant === tagVariants.BLUE && styles.blue]}
            id={id}
            onClick={onClick}
        >
            {icon && <Icon css={styles.icon} icon={icon} />}
            {children}
        </div>
    )
}

Tag.displayName = 'Tag'
Tag.defaultProps = defaultProps

export default Tag
