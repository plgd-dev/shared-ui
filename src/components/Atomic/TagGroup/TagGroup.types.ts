import { ReactNode } from 'react'
import { justifyContent } from './constants'

export type TagGroupJustifyContentType = (typeof justifyContent)[keyof typeof justifyContent]

export type Props = {
    children: ReactNode[] | ReactNode
    className?: string
    id?: string
    i18n: {
        more: string
        types: string
    }
    justifyContent?: TagGroupJustifyContentType
}

export const defaultProps: Partial<Props> = {
    justifyContent: justifyContent.START,
}
