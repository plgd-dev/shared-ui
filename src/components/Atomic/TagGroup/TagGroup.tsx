import React, { Children, FC, useCallback, useEffect, useRef, useState } from 'react'
import debounce from 'lodash/debounce'

import { Props, defaultProps } from './TagGroup.types'
import * as styles from './TagGroup.styles'
import Tag from '../Tag'
import { tagVariants } from '../Tag/constants'
import Modal from '../Modal'
import { justifyContent } from './constants'

const TagGroup: FC<Props> = (props) => {
    const { className, id, i18n, children, justifyContent: justifyContentProp } = props
    const [show, setShow] = useState(false)
    const [tagsToDisplay, setTagsToDisplay] = useState(0)
    const childrenArray = Children.toArray(children)

    const parent = useRef<any>(null)
    const tags: any = useRef<any>([])
    const moreTags: any = useRef<any>(null)

    const Inner = () => {
        if (childrenArray.length > tagsToDisplay) {
            return (
                <div css={[styles.tagGroup, justifyContentProp === justifyContent.END && styles.justifyEnd]}>
                    {childrenArray.slice(0, tagsToDisplay).map((item, key) => (
                        <div css={styles.tag} key={key}>
                            {item}
                        </div>
                    ))}
                    <div css={styles.tag}>
                        <Tag onClick={() => setShow(true)} variant={tagVariants.BLUE}>
                            {childrenArray.length - tagsToDisplay} more
                        </Tag>
                    </div>
                </div>
            )
        } else {
            return (
                <div css={[styles.tagGroup, justifyContentProp === justifyContent.END && styles.justifyEnd]}>
                    {childrenArray.map((tag, key) => (
                        <div css={styles.tag} key={key}>
                            {tag}
                        </div>
                    ))}
                </div>
            )
        }
    }

    const renderBody = () => (
        <div css={styles.tagGroup}>
            {childrenArray.map((tag, key) => (
                <div css={styles.tag} key={key}>
                    {tag}
                </div>
            ))}
        </div>
    )

    const calculate = useCallback(() => {
        const parentWidth = parent?.current?.clientWidth ?? 0
        let totalChildWidth = 2
        const childArray: number[] = []

        tags.current.forEach((tag: any, key: number) => {
            totalChildWidth += tag?.clientWidth
            if (key !== 0) {
                totalChildWidth += 4
            }
            childArray.push(tag?.clientWidth + 4)
        })

        if (totalChildWidth > parentWidth) {
            let currentWidth = moreTags?.current?.clientWidth ?? 0
            let numbersToRender = 0
            let findingIsOver = false

            childArray.forEach((child: any) => {
                if (currentWidth + child < parentWidth && !findingIsOver) {
                    currentWidth += child
                    numbersToRender++
                } else {
                    findingIsOver = true
                }
            })

            setTagsToDisplay(numbersToRender)
        } else {
            setTagsToDisplay(childArray.length)
        }
    }, [parent, tags, moreTags])

    const debouncedCallback = debounce(() => {
        calculate()
    }, 500)

    useEffect(() => {
        window.addEventListener('resize', debouncedCallback)
        return () => window.removeEventListener('resize', debouncedCallback)
    }, [debouncedCallback])

    useEffect(() => {
        calculate()
    }, [calculate])

    return (
        <div className={className} css={styles.parent} id={id} ref={parent}>
            <div css={styles.test}>
                <div css={styles.testInner}>
                    {childrenArray.map((tag, key) => (
                        <div css={styles.tag} key={key} ref={(el) => (tags.current[key] = el)}>
                            {tag}
                        </div>
                    ))}
                    <Tag ref={moreTags} variant={tagVariants.BLUE}>
                        {childrenArray.length} {i18n.more}
                    </Tag>
                </div>
            </div>

            {tagsToDisplay > 0 && <Inner />}
            {tagsToDisplay === 0 && (
                <Tag onClick={() => setShow(true)} variant={tagVariants.BLUE}>
                    {childrenArray.length} {i18n.more}
                </Tag>
            )}

            <Modal
                appRoot={document.getElementById('root')}
                maxWidth={600}
                onClose={() => setShow(false)}
                portalTarget={document.getElementById('modal-root')}
                renderBody={renderBody}
                show={show}
                title={i18n.modalHeadline}
            />
        </div>
    )
}

TagGroup.displayName = 'TagGroup'
TagGroup.defaultProps = defaultProps

export default TagGroup
