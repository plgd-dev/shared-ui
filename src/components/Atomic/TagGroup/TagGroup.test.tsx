import { render } from '../../../../test/jest-wrapper'
import TagGroup from './TagGroup'
import Tag from '../Tag'

describe('TagGroup Component', () => {
    it('renders children when there is enough space', () => {
        const { getByText, asFragment } = render(
            <TagGroup i18n={{ more: 'more', modalHeadline: 'Modal Headline' }}>
                <Tag>Tag 1</Tag>
                <Tag>Tag 2</Tag>
            </TagGroup>
        )
        expect(getByText('Tag 1')).toBeInTheDocument()
        expect(getByText('Tag 2')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })
})
