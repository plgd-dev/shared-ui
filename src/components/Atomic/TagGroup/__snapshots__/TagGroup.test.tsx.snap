// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagGroup Component renders children when there is enough space 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100%;
  position: relative;
}

.emotion-1 {
  position: absolute;
  width: 1px;
  height: 1px;
  visibility: hidden;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #F6F7F9;
  border-radius: 4px;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 12px;
  line-height: 18px;
  color: #81868C;
  padding: 3px 8px;
  border: 1px solid transparent;
  white-space: nowrap;
}

.emotion-7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #F6F7F9;
  border-radius: 4px;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 12px;
  line-height: 18px;
  color: #81868C;
  padding: 3px 8px;
  border: 1px solid transparent;
  white-space: nowrap;
  color: #2261AE;
  background: rgba(135,206,242,0.2);
  border: 1px solid transparent;
}

.emotion-8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #F6F7F9;
  border-radius: 4px;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 12px;
  line-height: 18px;
  color: #81868C;
  padding: 3px 8px;
  border: 1px solid transparent;
  white-space: nowrap;
  cursor: pointer;
  color: #2261AE;
  background: rgba(135,206,242,0.2);
  border: 1px solid transparent;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          <div
            class="emotion-4"
          >
            Tag 1
          </div>
        </div>
        <div
          class="emotion-3"
        >
          <div
            class="emotion-4"
          >
            Tag 2
          </div>
        </div>
        <div
          class="emotion-7"
        >
          2 more
        </div>
      </div>
    </div>
    <div
      class="emotion-8"
    >
      2 more
    </div>
  </div>
</DocumentFragment>
`;
